pkgs/_macro_host/lib/src/macro_cache.dart

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:dart_model/dart_model.dart';
+import 'package:macro_service/macro_service.dart';
+
+import 'macro_host.dart';
+
+class MacroResultsCache {
+  final _cache = <_MacroResultsCacheKey, _MacroResultsCacheValue>{};
+
+  final QueryService queryService;
+
+  MacroResultsCache(this.queryService);
+
+  /// Adds an entry into the cache.
+  ///
+  /// The cache keys are a combined value of the [annotation], and [request].
+  ///
+  /// The caceh values are based on each query in [queryResults] and a digest of
+  /// the [Model] response for that query.
+  void cache({
+    required QualifiedName annotation,
+    required AugmentRequest request,
+    required Iterable<({Query query, Model response})> queryResults,
+    required AugmentResponse response,
+  }) {
+    _cache[(
+      annotation: annotation.asRecord,
+      target: request.target.asRecord,
+      phase: request.phase,
+    )] = (
+      queries: queryResults.map((q) => q.asCachedResult),
+      response: response
+    );
+  }
+
+  Future<AugmentResponse?> cachedResult(
```

Comment:
It took me a while to figure out what's going here, particularly rerunning of queries :)

It's hard to come up with a fully descriptive name for this method ... rerunQueriesAndReturnCachedResultIfResultsUnchanged? :) ... maybe it would make sense to split up so the rerunning of the queries is separate from the check?

That would also help address my other comment which is that it would be nice not to throw away the query results, presumably they should go into some other cache so the macro can use them.

Feel free to leave TODOs for any/all of these :)

Reply:
I added a doc comment to at least explain this method :)